core/src/test/java/org/apache/calcite/test/StreamTest.java

```diff
@@ -286,15 +286,17 @@ private static String schemaFor(String name, Class<? extends TableFactory> clazz
             + "      LogicalTableScan(table=[[STREAM_JOINS, PRODUCTS]])\n")
         .explainContains(""
             + "EnumerableCalc(expr#0..6=[{inputs}], proj#0..1=[{exprs}], SUPPLIERID=[$t6])\n"
-            + "  EnumerableHashJoin(condition=[=($4, $5)], joinType=[inner])\n"
-            + "    EnumerableCalc(expr#0..3=[{inputs}], expr#4=[CAST($t2):VARCHAR(32) NOT NULL], proj#0..4=[{exprs}])\n"
-            + "      EnumerableInterpreter\n"
-            + "        BindableTableScan(table=[[STREAM_JOINS, ORDERS, (STREAM)]])\n"
-            + "    EnumerableTableScan(table=[[STREAM_JOINS, PRODUCTS]])")
+            + "  EnumerableMergeJoin(condition=[=($4, $5)], joinType=[inner])\n"
+            + "    EnumerableSort(sort0=[$4], dir0=[ASC])\n"
+            + "      EnumerableCalc(expr#0..3=[{inputs}], expr#4=[CAST($t2):VARCHAR(32) NOT NULL], proj#0..4=[{exprs}])\n"
+            + "        EnumerableInterpreter\n"
+            + "          BindableTableScan(table=[[STREAM_JOINS, ORDERS, (STREAM)]])\n"
```

Comment:
There are some cases that changed from HashJoin to MergeJoin. Are they expected?

Reply:
The traitset can be both required and derived.

Reply:
I think so.

Reply:
I'm confused by the `MergeJoin` is better, is there any problem with the cost estimation ?

Reply:
The cost model is orthogonal with this change.

Reply:
Previously `MergeJoin` was not taken because it did not really try to sort inputs. In other words, `MergeJoin` was taken **only** in case both inputs were already pre-sorted (which was only happening for literal `VALUES`).

I guess now it "abstract-converts" non-sorted rels to the sorted state, so `MergeJoin` can really succeed.

@hsyuan , is it the case?

Reply:
@vlsi Correct.

Reply:
This change reduces the amount the abstract converters and before this patch, the `EnumerableMergeJoinRule` did try to convert the input convention, i guess it is because there are redundant converters there so the cost estimation is affected.

Reply:
If you turn abstract converter in master branch, it generates the same plan.

Reply:
Thanks, got your idea ~